CIPs/CIP-69/CIP-69.md

```diff
@@ -0,0 +1,108 @@
+---
+cip: 69
+title: Batched Anchor Data Structure
+author: Joel Thorstensson (@oed)
+discussions-to: https://github.com/ceramicnetwork/CIP/issues/69
+status: Last Call
+category: Standards
+type: Core
+created: 2020-11-02
+---
+
+## Simple Summary
+This CIP describes the data structure of a blockchain anchor that batches multiple document updates together. 
+
+
+## Abstract
+An anchor service accepts document update anchor request from clients and batches them together in a single merkle tree, which is then anchored into a specific blockchain. The anchor service also sorts all of the requests and creates a bloom filter which describes which documents where updated in this batch, which is also included in the blockchain transaction. This allows external observers to more effectively determine if a document which they are interested in was updated.
+
+
+## Motivation
+There are two main motivations for this CIP. Firstly having an agreed upon spec for the structure of the batched anchor merkle tree makes it easier for all consumers of the ceramic protocol to know what to expect when parsing an *anchor record*. Secondly the sorting of updates and the inclusion of a bloom filter allows indexing services to effectively determine if an update was made which is relevant, and find this update in the tree effectively.
+
+
+## Specification
+
+First lets define the basic IPLD data structures used.
+```ipldsch
+type MerkleNode [Link]
+
+type TreeMetaData struct {
+  numEntries Int
+  bloomType String
+  bloomFilter {String:Any}
+}
+```
+Here the *MerkleNode* is simply an array of CIDs, which is used as the main building block for the merkle tree. The *TreeMetaData* contains some metadata about the tree, namely the number of entries and a bloom filter. The `bloomType` is used to describe which specific type of bloom filter is being used and is there for upgradeability purposes.
+
+### Tree structure
+The structure of the merkle tree is very straight forward. Starting from the bottom each individual update is paired with another update and a *MerkleNode* is created which has the leftmost update in position `0`, and the other update at position `1`. *MerkleNodes* are then created on each level until we reach the root. The root *MerkleNode* has an additional entry at position `2` which simply points to a *TreeMetaData* object. In the figure below we see a simple tree with four leafs.
+![merkletree(1)](https://user-images.githubusercontent.com/3909429/97851426-a83bf900-1cf5-11eb-9d9f-e7b568138850.png)
+In this example an *IPLD Path* can be used to reach any of the leafs. For example, to reach *Update 3* we would use the path `<root-cid>/1/0`. Similarly we can always get the *TreeMetaData* using the path `<root-cid>/2`.
+
+The merkle tree should be constructed as a balanced tree. See the [implementation][https://github.com/ceramicnetwork/ceramic-anchor-service/blob/4615c814ed98bb17cba8902480d8f22bde5397a1/src/merkle/merkle-tree.ts#L46-L65] for reference.
```

Comment:
So reading that implementation, I don't actually think it produces a balanced tree.  I filed https://github.com/ceramicnetwork/ceramic-anchor-service/issues/138 describing the issue.

Reply:
Thanks, will have to update here once fixed!

Reply:
@stbrody can you point me to the updated code snippet so that i can update and merge this?

Reply:
https://github.com/ceramicnetwork/ceramic-anchor-service/blob/ced4ea9f8c70aa09e6c87e3c1b2de5bbdf505157/src/merkle/merkle-tree.ts#L26-L61